pkg/storage/stores/indexshipper/storage/cached_client.go

```diff
@@ -20,137 +20,279 @@ const (
+		return []client.StorageObject{}, tableNames, err
+	}
+
+	// common objects and list of users having objects in a table were requested
+	if len(ss) == 1 {
+		tableName := ss[0]
+		return c.listTable(ctx, tableName)
+	}
+
+	// user objects in a table were requested
+	tableName := ss[0]
+	userID := ss[1]
+
+	userObjects, err := c.listUserIndexInTable(ctx, tableName, userID)
+	return userObjects, []client.StorageCommonPrefix{}, err
+}
+
+func (c *cachedObjectClient) listTableNames(ctx context.Context) ([]client.StorageCommonPrefix, error) {
+	if time.Since(c.tableNamesCacheBuiltAt) >= cacheTimeout {
+		buildCacheOnce(&c.buildTableNamesCacheWg, c.buildTableNamesCacheChan, func() {
+			c.err = nil
+			c.err = c.buildTableNamesCache(ctx, false)
+		})
 	}
 
 	// wait for cache build operation to finish, if running
-	c.buildCacheWg.Wait()
+	c.buildTableNamesCacheWg.Wait()
 
 	if c.err != nil {
-		return nil, nil, c.err
+		return nil, c.err
 	}
 
 	c.tablesMtx.RLock()
 	defer c.tablesMtx.RUnlock()
 
-	// list of tables were requested
-	if prefix == "" {
-		return []client.StorageObject{}, c.tableNames, nil
+	return c.tableNames, nil
+}
+
+func (c *cachedObjectClient) listTable(ctx context.Context, tableName string) ([]client.StorageObject, []client.StorageCommonPrefix, error) {
+	tbl := c.getTable(tableName)
+	if tbl == nil {
+		return []client.StorageObject{}, []client.StorageCommonPrefix{}, nil
 	}
 
-	// common objects and list of users having objects in a table were requested
-	if len(ss) == 1 {
-		tableName := ss[0]
-		if table, ok := c.tables[tableName]; ok {
-			return table.commonObjects, table.userIDs, nil
-		}
+	if time.Since(tbl.cacheBuiltAt) >= cacheTimeout {
+		buildCacheOnce(&tbl.buildCacheWg, tbl.buildCacheChan, func() {
+			tbl.err = nil
+			tbl.err = tbl.buildCache(ctx, c.ObjectClient, false)
+		})
+	}
 
-		return []client.StorageObject{}, []client.StorageCommonPrefix{}, nil
+	// wait for cache build operation to finish, if running
+	tbl.buildCacheWg.Wait()
+
+	if tbl.err != nil {
+		return nil, nil, tbl.err
 	}
 
-	// user objects in a table were requested
-	tableName := ss[0]
-	table, ok := c.tables[tableName]
-	if !ok {
-		return []client.StorageObject{}, []client.StorageCommonPrefix{}, nil
+	return tbl.commonObjects, tbl.userIDs, nil
+}
+
+func (c *cachedObjectClient) listUserIndexInTable(ctx context.Context, tableName, userID string) ([]client.StorageObject, error) {
+	tbl := c.getTable(tableName)
+	if tbl == nil {
+		return []client.StorageObject{}, nil
 	}
 
-	userID := ss[1]
-	if objects, ok := table.userObjects[userID]; ok {
-		return objects, []client.StorageCommonPrefix{}, nil
+	if time.Since(tbl.cacheBuiltAt) >= cacheTimeout {
+		buildCacheOnce(&tbl.buildCacheWg, tbl.buildCacheChan, func() {
+			tbl.err = nil
+			tbl.err = tbl.buildCache(ctx, c.ObjectClient, false)
+		})
+	}
+
+	// wait for cache build operation to finish, if running
+	tbl.buildCacheWg.Wait()
+
+	if tbl.err != nil {
+		return nil, c.err
+	}
+
```

Comment:
objects within the table are protected using the mtx from what I understand, it's missing here and in `listTable`
```
t.mtx.Lock()
defer t.mtx.Unlock()
```

Reply:
I deliberately didn't keep it locked to avoid contention. The table objects can change when refreshing the table names cache, however, that code does not clear/modify any data on the table objects. However, I think it would be better to have a read lock to avoid any possible bugs with future changes. I will see if I can change it. Thanks for the comment!

Reply:
just to be on the same page :) I am referring to the mtx in `table` which we are [locking](https://github.com/grafana/loki/pull/9798/files#diff-3a3ba0cf92979e89dab84da8c9e882ff5c38d7b467c5ed1cbeefd434fcb3c98cR290) when updating the common and user objects in `buildCache`

not taking the lock on the table names cache in this path might be alright I think

Reply:
ohh, sorry, my bad. I missed it. Good catch!